### PR TITLE
okx, websocket: avoid duplicate equivalent subscription routing

### LINKED ISSUE #2136: OKX: Duplicate WebSocket subscriptions for the same channel when a pair is enabled across multiple asset types (e.g., spot and margin)
When a trading pair is enabled for multiple asset types that share the same instrument ID on OKX (e.g., the same pair in both spot and margin trading), the WebSocket subscription logic generates identical subscription requests for channels like tickers and orderbooks.

Previously, with a single WebSocket connection, resubscribing to an already-active channel was harmless and did not cause errors.

However, when scaling to multiple WebSocket connections (e.g., due to rate limits or multi-connection support), these duplicate identical subscriptions are sent across different connections. This leads to:
* Unnecessary resource consumption (multiple connections receiving the same data stream).
* Potential checksum validation failures for orderbook channels, as the client may receive mismatched or conflicting updates from duplicated feeds.

To resolve this, subscriptions should be deduplicated at the exchange level before allocation/distribution across connections, ensuring each unique channel is subscribed to only once regardless of asset type overlap.

See: https://github.com/thrasher-corp/gocryptotrader/pull/2109#discussion_r2612508210

### PR DESCRIPTION
**Pull request by thrasher-**
Fixes #2136.

OKX can treat some spot/margin websocket subscriptions as equivalent. When the same pair/channel is enabled across both asset types, GoCryptoTrader could create duplicate or no-op connection work and mis-track the logical subscription.

This change:
- tracks equivalent logical subscriptions on existing connections before creating new ones
- attaches the logical subscription to the connection that already owns the inverse subscription
- runs the tracking pass before generic capacity-based routing so equivalent subscriptions are not misrouted to an unrelated connection with spare capacity
- guards tracked-subscription over-capacity handling in websocket subscription routing
- adds regression coverage for owning-connection selection, pre-routing tracking, misrouting prevention, and over-capacity handling

Tests:
- `go test ./exchange/websocket -run 'Test(ScaleConnectionsToSubscriptions|ConnectTracksOnExistingConnectionBeforeNewConnection|FlushChannels)$' -count=1`
- `go test ./exchanges/okx -run 'TestTrackEquivalent|TestGetSpotMarginEvaluator' -count=1`

### CODE REVIEW

**Comment by thrasher-** [comment]
@codex please review

File: exchanges/okx/ws_subscriptions.go [hunk 1]
```diff
@@ -85,3 +87,48 @@ incoming:
 	}
 	return eval
 }
+
+func inverseSpotMarginSubscription(sub *subscription.Subscription) (*subscription.Subscription, bool) {
```
Comment (gloriousCode) [inline exchanges/okx/ws_subscriptions.go:93]: Add test

File: exchange/websocket/subscriptions.go [hunk 2]
```diff
@@ -354,6 +354,24 @@ func (m *Manager) updateChannelSubscriptions(ctx context.Context, store *subscri
 	return nil
 }
 
+func (m *Manager) trackOnExistingConnection(ctx context.Context, ws *websocket, subs subscription.List) (subscription.List, error) {
```
Comment (gloriousCode) [inline exchange/websocket/subscriptions.go]: Add dedicated test. I think as a rule, Jarvis should add dedicated function tests for all functions added, even unexported unless its only a tiny obvious function eg `func getOne() int { return 1 }`

File: exchanges/okx/ws_subscriptions_test.go [hunk 3]
```diff
@@ -101,3 +105,67 @@ func TestNeedsOutboundSubscription(t *testing.T) {
...
+
+	remaining, err := tracked.trackEquivalentSubscriptionsOnExistingConnection(t.Context(), existingConn, subscription.List{spotSub})
+	require.NoError(t, err)
+	require.Empty(t, remaining, "equivalent spot subscription must be tracked on the existing connection")
+	require.Empty(t, existingConn.requests, "tracking on an existing connection must not emit a new outbound subscribe request")
+	require.NotNil(t, tracked.Websocket.GetSubscription(spotSub), "spot subscription must be tracked logically")
+	require.Len(t, existingConn.subscriptions.List(), 2, "existing connection store must track both logical subscriptions")
+}
+
+func TestTrackEquivalentSkipsConnectionWithoutInverse(t *testing.T) {
```
Comment (gloriousCode) [inline exchanges/okx/ws_subscriptions_test.go]: This looks like a t.Run scenario for testTrackEquivalentSubscriptionsOnExistingConnection

File: exchange/websocket/subscriptions.go [hunk 4]
```diff
@@ -354,6 +354,24 @@ func (m *Manager) updateChannelSubscriptions(ctx context.Context, store *subscri
...
+	}
+	remaining := subs
+	for _, conn := range ws.connections {
+		var err error
+		remaining, err = ws.setup.TrackOnExistingConnection(ctx, conn, remaining)
+		if err != nil {
+			return nil, err
+		}
+		if len(remaining) == 0 {
+			return nil, nil
```
[the remark this answers is not in the document]
Comment (thrasher-) [reply exchange/websocket/subscriptions.go]: Jarvis: Good point. Current callers only rely on len(...) == 0, so nil vs empty is behaviourally equivalent here. I left this unchanged to keep the fix scoped to the correctness issues, but I can fold in returning an empty non-nil slice as a small follow-up if needed.

File: exchange/websocket/manager.go [hunk 5]
```diff
@@ -528,15 +528,26 @@ func (m *Manager) connect(ctx context.Context) error {
...
-			if err := m.createConnectAndSubscribe(ctx, m.connectionManager[i], batchedSubs); err != nil {
+			toConnect, err := m.trackOnExistingConnection(ctx, m.connectionManager[i], batchedSubs)
+			if err != nil {
+				subscriptionError = common.AppendError(subscriptionError, fmt.Errorf("subscription error on [conn:%d] [URL:%s]: %w ", i+1, m.connectionManager[i].setup.URL, err))
+				continue
+			}
+			if len(toConnect) == 0 {
+				if m.verbose {
+					log.Debugf(log.WebsocketMgr, "%s websocket: [URL:%s] tracked logical subscriptions on existing connection. [Total Subs: %d] [Tracked: %d]", m.exchangeName, m.connectionManager[i].setup.URL, len(subs), len(batchedSubs))
+				}
```
[the remark this answers is not in the document]
Comment (thrasher-) [reply exchange/websocket/manager.go:593]: Jarvis: Fixed in https://github.com/thrasher-/gocryptotrader/commit/203a42d58dda2a15567ae23712dbc70faf204e02 . The multi-connection Connect() path now validates that tracked-only batches were actually recorded and surfaces ErrSubscriptionFailure/ErrSubscriptionsNotAdded if they were not. I also added a regression test covering the tracked-only/no-recording case.

File: exchange/websocket/manager.go [hunk 6]
```diff
@@ -528,15 +528,26 @@ func (m *Manager) connect(ctx context.Context) error {
...
+				continue
+			}
+			if err := m.createConnectAndSubscribe(ctx, m.connectionManager[i], toConnect); err != nil {
 				if errors.Is(err, common.ErrFatal) {
 					multiConnectFatalError = fmt.Errorf("cannot connect to [conn:%d] [URL:%s]: %w ", i+1, m.connectionManager[i].setup.URL, err)
 					break
 				}
 				subscriptionError = common.AppendError(subscriptionError, fmt.Errorf("subscription error on [conn:%d] [URL:%s]: %w ", i+1, m.connectionManager[i].setup.URL, err))
 			}
 			if m.verbose {
```
[the remark this answers is not in the document]
Comment (thrasher-) [reply exchange/websocket/manager.go:604]: Jarvis: Fixed in https://github.com/thrasher-/gocryptotrader/commit/203a42d58dda2a15567ae23712dbc70faf204e02 . The verbose connected log now only runs when createConnectAndSubscribe succeeds, so non-fatal subscription errors no longer produce a misleading success message.

File: exchanges/okx/ws_subscriptions_test.go [hunk 7]
```diff
@@ -101,3 +105,67 @@ func TestNeedsOutboundSubscription(t *testing.T) {
...
+	requests      []WSSubscriptionInformationList
+}
+
+func (c *subscriptionRecorderConnection) SendJSONMessage(_ context.Context, _ request.EndpointLimit, payload any) error {
+	req, ok := payload.(WSSubscriptionInformationList)
+	if !ok {
+		return nil
+	}
+	c.requests = append(c.requests, req)
+	return nil
```
[the remark this answers is not in the document]
Comment (thrasher-) [reply exchanges/okx/ws_subscriptions_test.go:123]: Jarvis: I left this helper as-is for now. The test already asserts on the recorded request slice, so an unexpected payload type still leaves the request list empty and fails the behaviour we care about. Happy to make the recorder stricter in a follow-up if needed.

File: exchanges/okx/ws_subscriptions.go (hunk 1 above)
Comment (thrasher-) [reply exchanges/okx/ws_subscriptions.go:93]: Jarvis: Added regression coverage in https://github.com/thrasher-/gocryptotrader/commit/99c55689be70af5df8a4101df61ecabed5bb9162 for tracking equivalent subscriptions on the owning connection and for skipping a connection that does not own the inverse subscription.

File: exchange/websocket/subscriptions.go (hunk 2 above)
Comment (thrasher-) [reply exchange/websocket/subscriptions.go]: Jarvis: Added dedicated unit tests in https://github.com/thrasher-/gocryptotrader/commit/99c55689be70af5df8a4101df61ecabed5bb9162 for trackOnExistingConnection itself, covering passthrough behaviour, multi-connection tracking until empty, and error propagation.

File: exchanges/okx/ws_subscriptions_test.go (hunk 3 above)
Comment (thrasher-) [reply exchanges/okx/ws_subscriptions_test.go]: Jarvis: Refactored this into t.Run subtests in https://github.com/thrasher-/gocryptotrader/commit/99c55689be70af5df8a4101df61ecabed5bb9162 so the owning-connection and wrong-connection cases now sit under the same test with clearer isolation.

File: exchanges/okx/ws_subscriptions.go (hunk 1 above)
Comment (gloriousCode) [reply exchanges/okx/ws_subscriptions.go:93]: Add a dedicated test, _please_

**Comment by thrasher-** [comment]
Jarvis: rebased onto current `master` and addressed the follow-up review items: OKX equivalent-subscription routing, tracked-only batch validation, dedicated tracked-subscription coverage, and websocket teardown cleanup for the macOS CI flake. The websocket package is now passing on macOS; the remaining failures appear to be unrelated live exchange/API flakiness.

**Comment by thrasher-** [comment]
Jarvis: the functional fix for Shazbert’s repro is in <https://github.com/thrasher-corp/gocryptotrader/commit/2df3855dc5cb80dd22657e098e0fe4ce41fdaf66>. That change refreshes the equivalent orderbook snapshot when an orderbook subscription is re-enabled and logically re-tracked onto an existing equivalent connection, so the margin/spot side does not resume from stale local state after disable → re-enable. For validation, I added a focused regression test for that path, ran `go test ./exchanges/okx -count=1`, and then verified the exact `gctcli` repro end-to-end in a local throwaway setup with `MaxWebsocketSubscriptionsPerConnection=3`: start with `ETH-USDT` margin, `OKB-USDT` margin, `BTC-USDT` margin, and `BTC-USDT` spot enabled; disable `BTC-USDT` margin; re-enable `BTC-USDT` margin; and confirm the `BTC-USDT` spot/margin orderbooks resume cleanly with no checksum mismatch or `orderbook data integrity compromised` errors. The later commit was only a misc-check wording tidy-up.

File: exchange/websocket/manager.go [hunk 8]
```diff
@@ -1104,28 +1120,32 @@ func (m *Manager) observeConnection(ctx context.Context, t *time.Timer) (exit bo
...
-			log.Warnf(log.WebsocketMgr, "%v websocket: has not received a traffic alert in %v. Reconnecting", m.exchangeName, timeout)
+			log.Warnf(log.WebsocketMgr, "%v websocket: has not received a traffic alert in %v. Reconnecting", m.exchangeName, m.trafficTimeout)
 		}
-		if m.IsConnected() {
-			go func() { // Without this the m.Shutdown() call below will deadlock
-				if err := m.Shutdown(); err != nil {
-					log.Errorf(log.WebsocketMgr, "%v websocket: trafficMonitor shutdown err: %s", m.exchangeName, err)
-				}
-			}()
+		if m.IsConnected() && onTimeout != nil {
```
Comment (shazbert) [inline exchange/websocket/manager.go:1156]: On timeout should always be non-nil and the function is unexported so the != nil check can be removed.

File: exchange/websocket/manager.go [hunk 9]
```diff
@@ -569,15 +569,31 @@ func (m *Manager) connect(ctx context.Context) error {
 		}
 
 		for _, batchedSubs := range common.Batch(subs, m.MaxSubscriptionsPerConnection) {
-			if err := m.createConnectAndSubscribe(ctx, ws, batchedSubs); err != nil {
+			toConnect, err := m.trackOnExistingConnection(ctx, ws, batchedSubs)
```
Comment (shazbert) [inline exchange/websocket/manager.go]: I think it might be better to call this above the common.Batch for loop because for example we might have a limit of 30 subs per connection and if we have 300 subs we could be reducing due to margin/spot and so on aggregate there are going to be connections that have sub 30 and across many managed connections we could in fact have more connections than what is needed. 

We could probably call the variable it returns something different like, reducedSubs.

File: exchange/websocket/subscriptions.go [hunk 10]
```diff
@@ -503,7 +535,7 @@ func (m *Manager) subscribeToConnection(ctx context.Context, conn Connection, su
 	}
 
 	usedCap := store.Len()
-	if m.MaxSubscriptionsPerConnection > 0 && usedCap == m.MaxSubscriptionsPerConnection {
+	if m.MaxSubscriptionsPerConnection > 0 && usedCap >= m.MaxSubscriptionsPerConnection {
```
Comment (shazbert) [inline exchange/websocket/subscriptions.go:587]: If used cap is above the m.MaxSubscriptionsPerConnection, I think we should error. Something is definitely wrong.

**Review by shazbert — CHANGES_REQUESTED** [review]
Tested the changes and they work perfectly. Just some minor nits left over then looks good to go.

File: exchange/websocket/manager.go (hunk 8 above)
Comment (thrasher-) [reply exchange/websocket/manager.go:1156]: Jarvis: I left this one as-is. The nil guard is harmless here, and observeTraffic is also exercised directly in tests with a nil callback, so keeping the check preserves that internal/test-time flexibility without changing behaviour.

File: exchange/websocket/manager.go (hunk 9 above)
Comment (thrasher-) [reply exchange/websocket/manager.go]: Jarvis: Agreed — This is now addressed in <https://github.com/thrasher-corp/gocryptotrader/commit/a8a26c017aa3b38d76f3c8de5cad3c772f15e213>. Connect() now reduces tracked subscriptions against existing connections before batching, while still re-checking each reduced batch later in the loop so newly created connections can absorb additional tracked subscriptions. I also added direct regression coverage for both the happy path and the pre-batch tracked-subscriptions-not-recorded failure path.

File: exchange/websocket/subscriptions.go (hunk 10 above)
Comment (thrasher-) [reply exchange/websocket/subscriptions.go:587]: Jarvis: Good point in the generic case, but a hard error on `usedCap > m.MaxSubscriptionsPerConnection` is not correct here. This store is tracking logical subscriptions, and for the OKX spot/margin-equivalent case one connection can legitimately hold more logical subscriptions than physical outbound subscription slots. Hard-erroring on the logical count would reject a valid configuration. The safer rule is that TrackOnExistingConnection must only suppress subscriptions that do not consume extra wire capacity; anything that still costs a real outbound slot must be returned in the remaining list (or use separate physical-capacity accounting).

File: exchange/websocket/manager.go [hunk 11]
```diff
@@ -568,16 +568,56 @@ func (m *Manager) connect(ctx context.Context) error {
...
 		}
 
-		for _, batchedSubs := range common.Batch(subs, m.MaxSubscriptionsPerConnection) {
-			if err := m.createConnectAndSubscribe(ctx, ws, batchedSubs); err != nil {
+		reducedSubs, err := m.trackOnExistingConnection(ctx, ws, subs)
+		if err != nil {
+			subscriptionError = common.AppendError(subscriptionError, fmt.Errorf("subscription error on [conn:%d] [URL:%s]: %w ", i+1, ws.setup.URL, err))
+			continue
+		}
+		if len(reducedSubs) != len(subs) {
```
Comment (gloriousCode) [inline exchange/websocket/manager.go]: Jarvis, could you please start adding replication steps to issues like these? This is a lot of convoluted code that I'm finding difficult to trigger. I can force it with code changes, but I'd rather test the real issue

File: exchanges/okx/ws_subscriptions.go (hunk 1 above)
Comment (thrasher-) [reply exchanges/okx/ws_subscriptions.go:93]: Jarvis: Added dedicated coverage in https://github.com/thrasher-/gocryptotrader/commit/ae0ce5ebb6c86a726c5ae90598d60c25c5c40f1a — there is now a direct TestInverseSpotMarginSubscription covering nil input, spot↔margin inversion, and non-spot/margin rejection rather than relying only on incidental coverage through the higher-level websocket tests.

File: exchange/websocket/manager.go (hunk 11 above)
Comment (thrasher-) [reply exchange/websocket/manager.go]: Jarvis: Fair ask. The underlying bug here was the OKX equivalent-subscription path, but part of the test follow-up has been to isolate the heavier websocket coverage into smaller direct tests (and use t.Run where it helps keep related cases focused) instead of relying only on the larger parallel websocket flows. I’ll be more explicit in future issue writeups about the concrete trigger/repro steps so the failure mode is easier to exercise and validate.

Comment (gloriousCode) [reply exchange/websocket/manager.go]: I will ask more directly, what do I need to configure and to what values, so that this updated code gets hit? I appreciate the tests, but I want to run the same scenario against master branch

Comment (thrasher-) [reply exchange/websocket/manager.go]: Jarvis: I dug back through the failing rebased CI runs to answer this more directly. There is not a single runtime config value to set on `master` that makes this path fire reliably, because the failure mode was not a simple config-gated production branch. The evidence points to **websocket test lifecycle / teardown sensitivity under package parallelism**, with the `macos-latest` arm64 runner being the environment that exposed it most consistently.

What I checked:
- the failing rebased runs for:
  - `a7cb271` → run `23235288307`
  - `46b31e1` → run `23235460884`
  - `ec843fa` → run `23236386224`
  - `ff84e2e` → run `23240819042`
- the corresponding diffs for:
  - `test: reduce websocket package parallelism`
  - `test: preserve safe websocket parallelism`
  - `test: refactor websocket flake coverage for parallelism`
- the full failed macOS job logs, including the goroutine dumps and timeout output

What the logs show:

1. `a7cb271` / run `23235288307`
- The websocket-related failure was isolated to `macos-latest` arm64; the other main OS jobs were green.
- The failures were not normal assertion mismatches. They were teardown/liveness failures:
  - `TestFlushChannels` failed with:
    - `Condition never satisfied`
    - `manager cleanup must wait for monitor goroutines`
  - `TestMonitorTraffic` failed with:
    - `Condition never satisfied`
- The package then hit the long-run kill:
  - `*** Test killed with quit: ran too long (11m0s)`
  - `FAIL github.com/thrasher-corp/gocryptotrader/exchange/websocket`
- The goroutine dump still had websocket mock upgrader goroutines alive from `TestScaleConnectionsToSubscriptions` while the package was being torn down.

That is a strong sign that the package was not failing on business logic; it was failing because websocket mock connections / monitor goroutines were not draining fast enough under that runner’s timing.

2. `46b31e1` / run `23235460884`
- macOS arm64 failed again.
- This time the package timed out after 10 minutes:
  - `panic: test timed out after 10m0s`
- The tests still running at timeout were:
  - `TestConnectionMessageErrors`
  - `TestExportedManagedSubscriptionFunctionsConcurrent/AddSubscriptions`
- The goroutine dump showed:
  - `(*Manager).shutdown()` blocked in `WaitGroup.Wait` at `manager.go:778`
  - websocket mock upgrader goroutines from `TestScaleConnectionsToSubscriptions` still alive
  - websocket monitor goroutines still active

So the package was effectively waiting on asynchronous websocket cleanup that had not completed yet. Again, that points to teardown / scheduling sensitivity rather than a platform-specific implementation bug.

3. `ec843fa` / run `23236386224`
- macOS arm64 went green immediately after the change set that reduced websocket-package parallelism in the sensitive spots.
- The key changes there were:
  - removing `t.Parallel()` from `TestConnectionMessageErrors`
  - removing `t.Parallel()` from `TestMonitorTraffic`
  - widening the `TestMonitorTraffic` `Eventually(...)` window back out to `5s` / `20ms`
  - keeping only the clearly isolated websocket parallelism, e.g. `TestTrackOnExistingConnection` and `TestConnectTrackOnExistingConnectionRequiresTrackedSubscriptions`

That is important because this commit did **not** fix a macOS-only production code path. It changed the **test scheduling / teardown pressure**. macOS passing immediately after that is strong evidence that the original problem was in the test shape, not in OS-specific websocket functionality.

4. `ff84e2e` / run `23240819042`
- When broader websocket parallelism/coverage was reintroduced, macOS arm64 failed again.
- The failure signatures were the same class of issue:
  - `httptest.Server blocked in Close after 5 seconds`
  - `use of closed network connection`
  - `i/o timeout`
  - `TestScaleConnectionsToSubscriptions` failing during websocket upgrade/write
  - `TestConnectionMessageErrors` failing cleanup assertions such as:
    - `manager cleanup must wait for monitor goroutines`
    - `connection monitor must stop during cleanup`

This is probably the strongest single piece of evidence, because the regression followed the **test scheduling / cleanup shape** rather than any obvious platform-specific behaviour in the production websocket manager.

So my conclusion is:

- the original macOS failures were real
- they were caused by **websocket test teardown/parallelism sensitivity**
- the most obvious pressure points were:
  - `TestConnectionMessageErrors`
  - `TestMonitorTraffic`
  - `TestScaleConnectionsToSubscriptions`
- the `macos-latest` arm64 runner exposed that sensitivity first because it was less forgiving to the timing assumptions around:
  - `httptest.Server` closure
  - gorilla/websocket connection shutdown
  - monitor goroutine drain timing
  - repeated `Connect()` / `Shutdown()` transitions under package-level parallel load

In other words, Linux was not proving those tests were safe; it was just staying on the right side of the cleanup/timing window.

There were a couple of unrelated failures in some of those runs on other jobs as well (for example a sonic job and a separate Ubuntu ARM exchange failure), but the websocket regression pattern itself was consistently tied to the macOS arm64 job.

For completeness, I also stress-ran the **current** websocket package state locally on Linux and could not reproduce the issue, including repeated targeted runs, race runs, shuffled full-package runs, and CPU-variation runs. That is consistent with the broader websocket lifecycle hardening that landed later, but it does **not** invalidate what the original macOS logs were showing at the time. It just means the current code is in a much better state now than those rebased failing snapshots were.

Comment (gloriousCode) [reply exchange/websocket/manager.go]: I am pleased that test issues have been resolved, but the issue being solved was not from tests. I will just take time to work it out

Comment (thrasher-) [reply exchange/websocket/manager.go]: Jarvis: Agreed — on this specific `if len(reducedSubs) != len(subs)` check, this is guarding the **functional pre-batch tracking path**, not the earlier test flake.

There are 3 outcomes after `trackOnExistingConnection(...)` runs:

- `len(reducedSubs) == len(subs)`
  - nothing was tracked before batching
- `len(reducedSubs) == 0`
  - everything was tracked before batching
- `len(reducedSubs) != len(subs)`
  - mixed case: some subscriptions were tracked onto existing connections before batching, and some still need the normal connect+subscribe path

This block is for that third case.

Once subscriptions are removed from `subs` before `common.Batch(...)`, the later batching loop never sees them again. So if an exchange implementation returns a reduced list but fails to record the pre-tracked subset in `ws.subscriptions`, `Connect()` can continue with incomplete logical subscription state.

That is why this reconstructs `trackedBeforeBatching` and validates:

- `ws.subscriptions.Missing(trackedBeforeBatching)`

If anything is missing, it fails immediately with `ErrSubscriptionFailure` / `ErrSubscriptionsNotAdded` instead of letting incomplete state leak forward into the OKX disable → re-enable path.

I also reproduced the functional master scenario locally by lowering OKX `MaxWebsocketSubscriptionsPerConnection` from `30` to `3`, then running the disable → enable flow for `BTC-USDT` margin with the equivalent spot/margin setup. On `master`, that re-enable path immediately hit:

- `orderbook data integrity compromised`
- `Okx BTC-USDT margin checksum mismatch`

So this validation is protecting a real functional gap, not just test scaffolding. I also added direct regression coverage for the missing-recorded-state branch in `TestConnectPreBatchTrackedSubscriptionsRequireRecordedState`.

File: exchange/websocket/connection.go [hunk 12]
```diff
@@ -84,6 +84,11 @@ type ConnectionSetup struct {
 	// messages based on the exchange's websocket server requirements to
```
Comment (gloriousCode) [inline exchange/websocket/connection.go:84]: Mr Jarvis sir, I have taken a look over of the code and have the following diff for you: https://gist.github.com/gloriousCode/67d2a4e4871d50dbc89bee645c75601b

I mostly found the flow to be a bit off, so it's why I wanted to make these changes. Would you consider applying it? I have an AI summary to try and back me up:

## What Changed

### 1. `TrackOnExistingConnection` contract updated

- **Before:** returned only `remaining` subscriptions.
- **Now:** returns `(remaining, tracked, err)`.

This makes tracked outcomes explicit and lets manager code apply/validate state centrally.

### 2. Manager now owns tracked state application/validation

Added helper functions in websocket manager flow:

- `applyTrackedSubscriptions(...)`
- `absorbTrackedSubscriptions(...)`
- `absorbTrackableSubscriptionsAndValidate(...)`

Also added comments for introduced helpers (including unexported).

### 3. `connect()` flow simplified and clarified

- Added explicit **pre-pass absorb** before batching:
  - absorbs trackable logical subs onto existing managed connections before capacity routing.
- Retains per-batch absorb check for later batches.
- Removed wasteful tracked-diff work in branches where remaining is already known to be zero.
- Naming cleanups for readability:
  - `remainingSubs`
  - `batchCandidates`
  - `batchRemainingSubs`

### 4. OKX hook adapted to new contract

`trackEquivalentSubscriptionsOnExistingConnection` now returns tracked subs instead of directly mutating manager/connection stores; manager performs the state updates.

### 5. Test coverage strengthened

Updated existing tests and added dedicated tests for introduced helper behavior, including error paths:

- `TestApplyTrackedSubscriptions`
- `TestAbsorbTrackableSubscriptionsAndValidate`
- existing absorb/connect tests updated for new return contract
- OKX tests updated for new expectations

## Why This Is Better

- **Clearer ownership:** exchange hook decides *what* is trackable; manager owns *how* state is recorded/validated.
- **Less hidden side effect risk:** no longer relies on hook implementations manually updating multiple stores correctly.
- **Centralized invariant checks:** tracked subs are validated in one manager path.
- **Cleaner connect logic:** less duplication, clearer control flow, better variable names.
- **Better testability:** helper functions now have direct tests, including failure modes.

## Validation

Ran:

```bash
go test ./exchange/websocket ./exchanges/okx -count=1

Comment (thrasher-) [reply exchange/websocket/connection.go:84]: Jarvis: Thanks — I’ve applied this approach.

Returning `(remaining, tracked, err)` from `TrackOnExistingConnection` makes the contract much clearer and lets the manager own tracked-subscription bookkeeping instead of inferring it afterwards.

I updated the manager to record tracked subscriptions centrally in both the websocket and connection stores, adjusted the OKX equivalent-subscription path to return tracked subscriptions without mutating stores directly, and updated the websocket/OKX tests around that ownership model.

Local validation is green, including the touched package race tests, lint, and full `make check`.

File: exchange/websocket/manager.go (hunk 11 above)
Comment (thrasher-) [reply exchange/websocket/manager.go]: Jarvis: I verified this live via `gctcli` with the same reduced-connection setup that reproduces the failure on `master`.

I lowered OKX `MaxWebsocketSubscriptionsPerConnection` from `30` to `3`, ran OKX with:
- `BTC-USDT` spot
- `BTC-USDT` margin
- `ETH-USDT` margin
- `OKB-USDT` margin

Then ran:
- `pair disable --exchange okx --pairs BTC-USDT --asset margin`
- `pair enable --exchange okx --pairs BTC-USDT --asset margin`

On `master`, that re-enable path reproduces:
- `Okx BTC-USDT margin checksum mismatch`
- `orderbook data integrity compromised`

On this branch, the same flow stayed clean:
- `BTC-USDT` margin subscriptions were removed and restored correctly
- spot + margin orderbooks resumed updating
- repeated `Okx passed checksum for pair BTC-USDT` messages continued after re-enable
- no checksum/integrity error surfaced

So this is now validated both by tests and by the live `gctcli` disable → enable repro.
